src/toolkits/drawing_classifier/drawing_classifier.cpp

```diff
@@ -289,18 +289,12 @@ void drawing_classifier::init_training(
     std::string feature_column_name, variant_type validation_data,
     std::map<std::string, flexible_type> opts) {
 
-  if (!(data.contains_column(feature_column_name))) {
-    log_and_throw(feature_column_name + "not found. Data passed in " + 
-      "does not contain the feature column model was trained with");
-  }
-
-  if (!(data.contains_column(target_column_name))) {
-    log_and_throw(target_column_name + "not found. Data passed in " + 
-      "does not contain the target column model was trained with");
-  }
+  ASSERT_MSG(data.contains_column(feature_column_name),
```

Comment:
Actually, `log_and_throw` is the correct mechanism for validating user inputs

`ASSERT_MSG` is for correcting programmer errors

Reply:
```
#define ASSERT_MSG(condition, fmt, ...)                                  \
  do {                                                                   \
    if (__builtin_expect(!(condition), 0)) {                             \
      auto throw_error = [&]() GL_GCC_ONLY(GL_COLD_NOINLINE_ERROR) {     \
        logstream(LOG_ERROR) << "Check failed: " << #condition << ":\n"; \
        std::ostringstream ss;                                           \
        ss << "Assertion Failure: " << #condition << ": " << fmt;        \
        logger(LOG_ERROR, fmt, ##__VA_ARGS__);                           \
        __print_back_trace();                                            \
        TURI_LOGGER_FAIL_METHOD(ss.str().c_str());                       \
        ASSERT_UNREACHABLE();                                            \
      };                                                                 \
      throw_error();                                                     \
      TURI_BUILTIN_UNREACHABLE();                                        \
    }                                                                    \
  } while (0)
```
The impl is same for log_and_throw. It doesn't really hault the program.

Reply:
```
#define log_and_throw(message)                        \
  do {                                                \
    auto throw_error = [&]() GL_COLD_NOINLINE_ERROR { \
      logstream(LOG_ERROR) << (message) << std::endl; \
      throw(std::string(message));                    \
    };                                                \
    throw_error();                                    \
  } while (0)
```
log_and_throw. Ah, slightly different. I would agree `ASSERT_MSG` is for programmers. Thanks for pointing out.

Reply:
I can't speak to the implementation differences, but maybe @hoytak can.

The intent is for `log_and_throw` to be user-facing. Assertions are for developers and can contain more technical information, such as backtraces. We also want to make it easy to tell at a glance what error strings are meant to be user-facing

Reply:
yeah, that's great. The reason I use asser_msg is that it's neater. Since they have different semantics, I should differentiate them.